File: pkg/sql/idxusage/index_usage_stats_rec.go
// Copyright 2022 The Cockroach Authors.
//
// Use of this software is governed by the CockroachDB Software License
// included in the /LICENSE file.

package idxusage

import (
	"fmt"
	"math"
	"time"

	"github.com/cockroachdb/cockroach/pkg/roachpb"
	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
	"github.com/cockroachdb/cockroach/pkg/settings"
	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
)

// IndexStatsRow is a wrapper type around
// serverpb.TableIndexStatsResponse_ExtendedCollectedIndexUsageStatistics that
// implements additional methods to support unused index recommendations and
// hold testing knobs.
type IndexStatsRow struct {
	TableID          roachpb.TableID
	IndexID          roachpb.IndexID
	CreatedAt        *time.Time
	LastRead         time.Time
	IndexType        string
	IsUnique         bool
	UnusedIndexKnobs *UnusedIndexRecommendationTestingKnobs
}

// defaultUnusedIndexDuration is a week.
const defaultUnusedIndexDuration = 7 * 24 * time.Hour

// DropUnusedIndexDuration registers the index unuse duration at which we
// begin to recommend dropping the index.
var DropUnusedIndexDuration = settings.RegisterDurationSetting(
	settings.ApplicationLevel,
	"sql.index_recommendation.drop_unused_duration",
	"the index unused duration at which we begin to recommend dropping the index",
	defaultUnusedIndexDuration,
	settings.NonNegativeDuration,
	settings.WithPublic,
)

const indexExceedUsageDurationReasonPlaceholder = "This index has not been used in over %sand can be removed for better write performance."
const indexNeverUsedReason = "This index has not been used and can be removed for better write performance."

// UnusedIndexRecommendationTestingKnobs provides hooks and knobs for unit tests.
type UnusedIndexRecommendationTestingKnobs struct {
	// GetCreatedAt allows tests to override the creation time of the index.
	GetCreatedAt func() *time.Time
	// GetLastRead allows tests to override the time the index was last read.
	GetLastRead func() time.Time
	// GetCurrentTime allows tests to override the current time.
	GetCurrentTime func() time.Time
}

// ModuleTestingKnobs implements base.ModuleTestingKnobs interface.
func (*UnusedIndexRecommendationTestingKnobs) ModuleTestingKnobs() {}

// GetRecommendationsFromIndexStats gets index recommendations from the given index
// if applicable.
func (i IndexStatsRow) GetRecommendationsFromIndexStats(
	dbName string, st *cluster.Settings,
) []*serverpb.IndexRecommendation {
	var recommendations = []*serverpb.IndexRecommendation{}
	// Omit fetching index recommendations for the 'system' database.
	if dbName == catconstants.SystemDatabaseName {
		return recommendations
	}
	rec := i.maybeAddUnusedIndexRecommendation(DropUnusedIndexDuration.Get(&st.SV))
	if rec != nil {
		recommendations = append(recommendations, rec)
	}
	return recommendations
}

func (i IndexStatsRow) maybeAddUnusedIndexRecommendation(
	unusedIndexDuration time.Duration,
) *serverpb.IndexRecommendation {
	if i.IsUnique {
		return nil
	}

	var rec *serverpb.IndexRecommendation

	if i.UnusedIndexKnobs == nil {
		rec = i.recommendDropUnusedIndex(timeutil.Now(), i.CreatedAt,
			i.LastRead, unusedIndexDuration)
	} else {
		rec = i.recommendDropUnusedIndex(i.UnusedIndexKnobs.GetCurrentTime(),
			i.UnusedIndexKnobs.GetCreatedAt(), i.UnusedIndexKnobs.GetLastRead(), unusedIndexDuration)
	}
	return rec
}

// recommendDropUnusedIndex checks whether the last usage of an index
// qualifies the index as unused, if so returns an index recommendation.
func (i IndexStatsRow) recommendDropUnusedIndex(
	currentTime time.Time,
	createdAt *time.Time,
	lastRead time.Time,
	unusedIndexDuration time.Duration,
) *serverpb.IndexRecommendation {
	lastActive := lastRead
	if lastActive.Equal(time.Time{}) && createdAt != nil {
		lastActive = *createdAt
	}
	// If we do not have the creation time and index has never been read. Recommend
	// dropping with a "never used" reason.
	if lastActive.Equal(time.Time{}) {
		return &serverpb.IndexRecommendation{
			TableID: i.TableID,
			IndexID: i.IndexID,
			Type:    serverpb.IndexRecommendation_DROP_UNUSED,
			Reason:  indexNeverUsedReason,
		}
	}
	// Last usage of the index exceeds the unused index duration.
	if currentTime.Sub(lastActive) >= unusedIndexDuration {
		return &serverpb.IndexRecommendation{
			TableID: i.TableID,
			IndexID: i.IndexID,
			Type:    serverpb.IndexRecommendation_DROP_UNUSED,
			Reason:  fmt.Sprintf(indexExceedUsageDurationReasonPlaceholder, formatDuration(unusedIndexDuration)),
		}
	}
	return nil
}

func formatDuration(d time.Duration) string {
	const numHoursInDay = 24
	const numMinutesInHour = 60
	const numSecondsInMinute = 60

	days := int64(d.Hours()) / (numHoursInDay)
	hours := int64(math.Floor(d.Hours())) % numHoursInDay
	minutes := int64(math.Floor(d.Minutes())) % numMinutesInHour
	seconds := int64(math.Floor(d.Seconds())) % numSecondsInMinute

	var daysSubstring string
	var hoursSubstring string
	var minutesSubstring string
	var secondsSubstring string

	if days > 0 {
		daysSubstring = fmt.Sprintf("%d days, ", days)
	}
	if hours > 0 {
		hoursSubstring = fmt.Sprintf("%d hours, ", hours)
	}
	if minutes > 0 {
		minutesSubstring = fmt.Sprintf("%d minutes, ", minutes)
	}
	if seconds > 0 {
		secondsSubstring = fmt.Sprintf("%d seconds, ", seconds)
	}

	return fmt.Sprintf("%s%s%s%s", daysSubstring, hoursSubstring, minutesSubstring, secondsSubstring)
}

File: pkg/ui/workspaces/cluster-ui/src/insights/schemaInsights/indexUsageStatsRec.ts
// Copyright 2022 The Cockroach Authors.
//
// Use of this software is governed by the CockroachDB Software License
// included in the /LICENSE file.

import moment from "moment-timezone";

export const indexNeverUsedReason =
  "This index has not been used and can be removed for better write performance.";

const minDate = moment.utc("0001-01-01"); // minimum value as per UTC.

type DropIndexRecommendation = {
  recommend: boolean;
  reason: string;
};

export interface IndexUsageStatistic {
  created_at?: string;
  last_read?: string;
  unused_threshold: string;
}

export function recommendDropUnusedIndex(
  indexUsageStat: IndexUsageStatistic,
): DropIndexRecommendation {
  const createdAt = indexUsageStat.created_at
    ? moment.utc(indexUsageStat.created_at)
    : minDate;
  const lastRead = indexUsageStat.last_read
    ? moment.utc(indexUsageStat.last_read)
    : minDate;
  let lastActive = createdAt;
  if (lastActive.isSame(minDate) && !lastRead.isSame(minDate)) {
    lastActive = lastRead;
  }

  if (lastActive.isSame(minDate)) {
    return { recommend: true, reason: indexNeverUsedReason };
  }

  const unusedThreshold = moment.duration(
    "PT" + indexUsageStat.unused_threshold.toUpperCase(),
  );
  return {
    recommend: true,
    reason: `This index has not been used in over ${formatMomentDuration(
      unusedThreshold,
    )} and can be removed for better write performance.`,
  };
}

export function formatMomentDuration(duration: moment.Duration): string {
  const numSecondsInMinute = 60;
  const numMinutesInHour = 60;
  const numHoursInDay = 24;

  const seconds = Math.floor(duration.as("s")) % numSecondsInMinute;
  const minutes = Math.floor(duration.as("m")) % numMinutesInHour;
  const hours = Math.floor(duration.as("h")) % numHoursInDay;
  const days = Math.floor(duration.as("d"));

  const daysSubstring = days > 0 ? `${days} days, ` : "";
  const hoursSubstring = hours > 0 ? `${hours} hours, ` : "";
  const minutesSubstring = minutes > 0 ? `${minutes} minutes, ` : "";
  const secondsSubstring = seconds > 0 ? `${seconds} seconds, ` : "";

  return `${daysSubstring}${hoursSubstring}${minutesSubstring}${secondsSubstring}`;
}
